src/ast/codegen_llvm.cpp

```diff
@@ -495,14 +499,15 @@ void CodegenLLVM::visit(Call &call)
     } else {
       b_.CreateStore(b_.getInt64(bpftrace_.strlen_), strlen);
     }
-    AllocaInst *buf = b_.CreateAllocaBPF(bpftrace_.strlen_, "str");
-    b_.CREATE_MEMSET(buf, b_.getInt8(0), bpftrace_.strlen_, 1);
     call.vargs->front()->accept(*this);
-    b_.CreateProbeReadStr(ctx_, buf, b_.CreateLoad(strlen), expr_, call.loc);
+    // zero it out first
+    b_.CreateProbeRead(ctx_, str_map, bpftrace_.strlen_,
+                      ConstantExpr::getCast(Instruction::IntToPtr, zeroed_area_ptr, b_.getInt8PtrTy()), call.loc);
```

Comment:
Would a memset work here instead? I see the intention is to clear out map entry before reading the string in but it feels a little funky.

I did a small test and it loaded ok on my system:
http://ix.io/2oRP
diffed on https://github.com/danobi/libbpf-rs

Reply:
The same question was asked of @mmarchini in his original RFC on storing printf args in a map: https://github.com/iovisor/bpftrace/pull/750#discussion_r292685014

> memset will only work for small values (even if we're using a map). I'm on another branch right now, I'll try to share more information later and I'll add a comment here explaining why this is implemented this way.

Unfortunately I don't know more detail than that! I'd definitely appreciate some detail from @mmarchini on this one.

Reply:
this is a wild guess, but IIRC I observed that the LLVM IR output for memset() was something like:

```
store a 0 at offset +16
store a 0 at offset +8
store a 0 at offset +0
```

so maybe you reach limits of BPF program size if you memset() a large enough value? whereas I think a call to probe_read() would do the work outside of our BPF program.

admittedly I'm not satisfied with that guess, since BPF programs are allowed to be pretty big nowadays.

Reply:
Maybe llvm unrolls the memset on small enough regions of memory. And maybe verifier doesn't like loop-based memsets.